### PR TITLE
[SPARK-44691][SQL][CONNECT] Move Subclasses of AnalysisException to sql/api

### PR DESCRIPTION
**Pull request by heyihong**
### What changes were proposed in this pull request?

- Move subclasses of AnalysisException to sql/api
- Refactor some dependent utilities

### Why are the changes needed?

- Supporting throws these exceptions for better compatibility with the existing control flow behaviors

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

- Existing tests

### CODE REVIEW

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/analysis/NoSuchItemException.scala
Comment (hvanhovell) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/analysis/NoSuchItemException.scala:1]: Convention: `noSuchItemsExceptions.scala` as the file name.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala
Comment (hvanhovell) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala:1]: Convention: rename this to `alreadyExistsExceptions.scala`.

File: sql/api/src/main/java/org/apache/spark/sql/connector/catalog/IdentifierImpl.java [hunk 1]
```diff
@@ -17,14 +17,14 @@
 
 package org.apache.spark.sql.connector.catalog;
 
+import org.apache.arrow.util.Preconditions;
```
Comment (amaliujia) [inline sql/api/src/main/java/org/apache/spark/sql/connector/catalog/IdentifierImpl.java:20]: Is this used?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 2]
```diff
@@ -100,13 +100,7 @@ object SparkStringUtils extends Logging {
...
   }
 
-  def quoteIdentifier(name: String): String = {
-    // Escapes back-ticks within the identifier name with double-back-ticks, and then quote the
-    // identifier with back-ticks.
-    "`" + name.replace("`", "``") + "`"
-  }
-
-  /**
+    /**
```
Comment (amaliujia) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:103]: revert this?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/QuotingUtils.scala [hunk 3]
```diff
@@ -37,6 +49,18 @@ object QuotingUtils {
     }
   }
 
+  def quoted(namespace: Array[String]): String = {
+    namespace.map(quoteIfNeeded).mkString(".")
+  }
+
+  def quoted(ident: Identifier): String = {
```
Comment (amaliujia) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/QuotingUtils.scala:56]: Nit: I am wondering if this should be renamed to `quotedIdentifier`. Same for the above which might be `quotedNamespace`?

**Review by amaliujia — COMMENTED** [review]
Some small comments.

File: sql/api/src/main/java/org/apache/spark/sql/connector/catalog/IdentifierImpl.java (hunk 1 above)
Comment (heyihong) [reply sql/api/src/main/java/org/apache/spark/sql/connector/catalog/IdentifierImpl.java:20]: Yes
```
  IdentifierImpl(String[] namespace, String name) {
    Preconditions.checkNotNull(namespace, "Identifier namespace cannot be null");
    Preconditions.checkNotNull(name, "Identifier name cannot be null");
    this.namespace = namespace;
    this.name = name;
  }
```

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/QuotingUtils.scala (hunk 3 above)
Comment (heyihong) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/QuotingUtils.scala:56]: It may be easier to use if quoted supports overload?

Comment (heyihong) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/QuotingUtils.scala:56]: `quoted` may be easier to use because of the overload support?

**Review by amaliujia — APPROVED** [review]

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/QuotingUtils.scala (hunk 3 above)
Comment (amaliujia) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/QuotingUtils.scala:56]: I am fine with either way

File: sql/api/src/main/java/org/apache/spark/sql/connector/catalog/Identifier.java
Comment (cloud-fan) [inline sql/api/src/main/java/org/apache/spark/sql/connector/catalog/Identifier.java:1]: It's weird to only move one DS v2 API to the api module. Shall we move all of them?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/analysis/alreadyExistException.scala [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.util.AttributeNameParser
+import org.apache.spark.sql.catalyst.util.QuotingUtils.{quoted, quoteIdentifier, quoteNameParts}
+import org.apache.spark.sql.connector.catalog.Identifier
+
+/**
+ * Thrown by a catalog when an item already exists. The analyzer will rethrow the exception
+ * as an [[org.apache.spark.sql.AnalysisException]] with the correct position information.
+ */
+class DatabaseAlreadyExistsException(db: String)
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/analysis/alreadyExistException.scala:30]: is there any actual change? or it's just code move around?

Comment (heyihong) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/analysis/alreadyExistException.scala:30]: Some dependent utilities in these exceptions are moved to sql/api as well for avoiding dependency on catalyst module. The change is reflected in the import as well
Before:
```
import org.apache.spark.sql.catalyst.util.{quoteIdentifier, quoteNameParts }
import org.apache.spark.sql.connector.catalog.CatalogV2Implicits.IdentifierHelper
```
After:
```
import org.apache.spark.sql.catalyst.util.QuotingUtils.{quoted, quoteIdentifier, quoteNameParts}
```

File: sql/api/src/main/java/org/apache/spark/sql/connector/catalog/Identifier.java
Comment (heyihong) [reply sql/api/src/main/java/org/apache/spark/sql/connector/catalog/Identifier.java:1]: It is a good point that we shall move all of them to the api module. But it seems to be out of scope for this pr. Ca we do this in a follow up?

Comment (cloud-fan) [reply sql/api/src/main/java/org/apache/spark/sql/connector/catalog/Identifier.java:1]: yea, can you create a JIRA ticket? to make sure we don't lose track.

**Review by cloud-fan — APPROVED** [review]

File: sql/api/src/main/java/org/apache/spark/sql/connector/catalog/Identifier.java
Comment (heyihong) [reply sql/api/src/main/java/org/apache/spark/sql/connector/catalog/Identifier.java:1]: Yes, https://issues.apache.org/jira/browse/SPARK-44744

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Pull request closed without merging**